src/services/device-management/iot-device.service.ts

```diff
@@ -384,89 +482,178 @@ export class IoTDeviceService {
+            const { deviceModelId } = map.iotDeviceDto;
+            if (!(deviceModelId in groups)) {
+                groups[deviceModelId] = [];
+                result.push(groups[deviceModelId]);
+            }
+            groups[deviceModelId].push(map);
+        });
+
+        return result;
     }
 
-    async mapDeviceModel(
-        iotDevice: IoTDevice,
-        createIoTDeviceDto: CreateIoTDeviceDto
-    ): Promise<DeviceModel> {
-        if (createIoTDeviceDto.deviceModelId == undefined) {
-            return null;
-        }
-        const deviceModel = await this.deviceModelService.getByIdWithRelations(
-            createIoTDeviceDto.deviceModelId
+    async mapDeviceModels(iotDevicesDtoMap: CreateIoTDeviceMapDto[]): Promise<void> {
+        const iotDevicesDtoByDeviceProfiles = this.groupByDeviceProfile(iotDevicesDtoMap);
+        const iotDevicesWithProfiles = _.flatMap(
+            // Check if the client sent a device profile. Since they're grouped by profile,
+            // the profile of the first device in the group is the same as the rest
+            iotDevicesDtoByDeviceProfiles.filter(
+                map => map[0]?.iotDeviceDto.deviceModelId
+            )
         );
-        const deviceModelOrganisationId = deviceModel.belongsTo.id;
-        const application = await this.applicationService.findOneWithOrganisation(
-            iotDevice.application.id
+        const deviceModels = await this.deviceModelService.getByIdsWithRelations(
+            iotDevicesWithProfiles.map(iotDevice => iotDevice.iotDeviceDto.deviceModelId)
+        );
+        const applications = await this.applicationService.findManyWithOrganisation(
+            iotDevicesWithProfiles.map(({ iotDevice }) => iotDevice.application.id)
         );
-        if (deviceModelOrganisationId != application.belongsTo.id) {
-            throw new BadRequestException(ErrorCodes.OrganizationDoesNotMatch);
-        }
 
-        return deviceModel;
-    }
+        for (const map of iotDevicesWithProfiles) {
+            const applicationMatch = applications.find(
+                application => application.id === map.iotDevice.application.id
+            );
 
-    private async setApplication(
-        createIoTDeviceDto: CreateIoTDeviceDto,
-        iotDevice: IoTDevice
-    ) {
-        if (createIoTDeviceDto.applicationId != null) {
-            iotDevice.application = await this.applicationService.findOneWithoutRelations(
-                createIoTDeviceDto.applicationId
+            if (!applicationMatch) {
+                map.error = {
+                    message: ErrorCodes.ApplicationDoesNotExist,
+                };
+                continue;
+            }
+
+            const deviceModelMatch = deviceModels.find(
+                model => model.id === map.iotDeviceDto.deviceModelId
             );
-        } else {
-            iotDevice.application = null;
+
+            if (deviceModelMatch?.belongsTo.id !== applicationMatch.belongsTo.id) {
```

Comment:
Verify that the DeviceModel actually exists here. Currently it throws an error about the organization not matching if an invalid DeviceModel has been chosen.